{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import math\n",
    "\n",
    "import sys, os\n",
    "sys.path.append(\"../\")\n",
    "\n",
    "from iotools.readOxford import ReadOxford\n",
    "from iotools.readrpkm import ReadRPKM\n",
    "from iotools.io_model import WriteModel\n",
    "from inference.linreg_association import LinRegAssociation\n",
    "from inference.empirical_bayes import EmpiricalBayes\n",
    "from utils import hyperparameters\n",
    "from inference import logmarglik\n",
    "from iotools import readgtf\n",
    "from utils import gtutils\n",
    "from utils import mfunc\n",
    "from utils.containers import ZstateInfo\n",
    "from utils.printstamp import printStamp\n",
    "\n",
    "from sklearn.preprocessing import scale\n",
    "\n",
    "home = '/home/fsimone'\n",
    "gtfpath = os.path.join(home,\"datasets/gtex/gencode.v19.annotation.gtf.gz\")\n",
    "chrom = 1\n",
    "samplepath = os.path.join(home, \"datasets/gtex/donor_ids.fam\")\n",
    "gtpath = os.path.join(home, \"datasets/gtex/GTEx_450Indiv_genot_imput_info04_maf01_HWEp1E6_dbSNP135IDs_donorIDs_dosage_chr\"+str(chrom)+\".gz\")\n",
    "rpkmpath = os.path.join(home, \"datasets/gtex/gtex_wholeblood_normalized.expression.txt\")\n",
    "params = [0.1, 0.0, 0.01, 0.001, 0.005]\n",
    "zmax = 1\n",
    "# for testing purposes I keep using the previous gene expression\n",
    "# rpkmpath = os.path.join(home, \"datasets/gtex/gtex_wholeblood_normalized.lm_corr_final2.exp.txt\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation (use complete gene name in gtf without trimming the version)\n",
    "gene_info = readgtf.gencode_v12(gtfpath, include_chrom = chrom, trim=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-02-27 17:19:35 - started reading genotype\n",
      "Read 892083 snps in 450 samples.\n",
      "2018-02-27 17:21:31 - Finished readings snps\n"
     ]
    }
   ],
   "source": [
    "# read Genotype\n",
    "oxf = ReadOxford(gtpath, samplepath, chrom, \"gtex\")\n",
    "genotype = np.array(oxf.dosage)\n",
    "samplenames = oxf.samplenames\n",
    "snps = oxf.snps_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quality control\n",
    "f_snps, f_genotype = gtutils.remove_low_maf(snps, genotype, 0.1)\n",
    "gt = gtutils.normalize(f_snps, f_genotype)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene Expression\n",
    "rpkm = ReadRPKM(rpkmpath, \"gtex\")\n",
    "expression = rpkm.expression\n",
    "expr_donors = rpkm.donor_ids\n",
    "gene_names = rpkm.gene_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-02-27 17:22:55 - Selection of samples\n"
     ]
    }
   ],
   "source": [
    "# Selection\n",
    "printStamp(\"Selection of samples\")\n",
    "vcfmask, exprmask = mfunc.select_donors(samplenames, expr_donors)\n",
    "genes, indices = mfunc.select_genes(gene_info, gene_names)\n",
    "\n",
    "\n",
    "min_snps = 200\n",
    "pval_cutoff = 0.001\n",
    "window = 1000000\n",
    "zmax = zmax    # z parameter\n",
    "init_params = np.array(params)\n",
    "init_params[4] = 1 / init_params[4] / init_params[4]\n",
    "\n",
    "model = WriteModel(\".\", chrom)\n",
    "\n",
    "batch_size = None\n",
    "gene_number = len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 70 high r2 genes\n",
      "\n",
      "Gene ENSG00000158825.5, CHR 1, R2 value: 0.3754060956719503 \n",
      "Gene ENSG00000183386.5, CHR 1, R2 value: 0.4036021829816217 \n",
      "Gene ENSG00000116157.5, CHR 1, R2 value: 0.4385918707779852 \n"
     ]
    }
   ],
   "source": [
    "# load gene ids of interest\n",
    "\n",
    "gene_list = []\n",
    "# genelistfile = \"target_gene_list\" # r2 > 0.1 and px > 2*gx\n",
    "genelistfile = \"genes4testing_highr2\" # r2 > 0.1\n",
    "r2val = []\n",
    "predixcan_r2val = []\n",
    "with open(genelistfile, 'r') as instream:\n",
    "    for line in instream:\n",
    "        arr = line.strip().split()\n",
    "        gene_list.append(arr[0])\n",
    "        r2val.append(arr[1])\n",
    "        predixcan_r2val.append(arr[2])\n",
    "\n",
    "print(\"Read {:d} high r2 genes\\n\".format(len(gene_list)))\n",
    "        \n",
    "selected_gene_ids = []\n",
    "for i in gene_info:\n",
    "    trimid = i.ensembl_id.split(\".\")[0]\n",
    "    if trimid in gene_list:\n",
    "        print(\"Gene {:s}, CHR {:d}, R2 value: {:s} \".format(i.ensembl_id, i.chrom, r2val[gene_list.index(trimid)]))\n",
    "        # print(\"R2 value: \",r2val[gene_list.index(trimid)], \"\\t\", predixcan_r2val[gene_list.index(trimid)])\n",
    "        selected_gene_ids.append(i.ensembl_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "342 GeneInfo(name='CDA', ensembl_id='ENSG00000158825.5', chrom=1, start=20915440, end=20945401)\n"
     ]
    }
   ],
   "source": [
    "for i, gene in enumerate(genes):\n",
    "    k = indices[i]\n",
    "    if gene.ensembl_id in selected_gene_ids[0]:\n",
    "        break\n",
    "print(k,gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 3096 SNPs, reduced to 200 SNPs (max p-value 0.0441457) for CDA\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# select only the cis-SNPs\n",
    "cismask = mfunc.select_snps(gene, f_snps, window)\n",
    "if len(cismask) > 0:\n",
    "    target = expression[k, exprmask]\n",
    "    target = scale(target, with_mean=True, with_std=True)\n",
    "    predictor = gt[cismask][:, vcfmask]\n",
    "    snpmask = cismask\n",
    "\n",
    "    # if number of cis SNPs > threshold, use p-value cut-off\n",
    "    if len(cismask) > min_snps:\n",
    "        assoc_model = LinRegAssociation(predictor, target, min_snps, pval_cutoff)\n",
    "        pvalmask = cismask[assoc_model.selected_variables]\n",
    "        print (\"Found {:d} SNPs, reduced to {:d} SNPs (max p-value {:g}) for {:s}\".format(len(cismask), len(pvalmask), assoc_model.ordered_pvals[len(pvalmask) - 1], gene.name))\n",
    "        predictor = gt[pvalmask][:, vcfmask]\n",
    "        snpmask = pvalmask\n",
    "    else:\n",
    "        print (\"Found {:d} SNPs for {:s}\".format(len(cismask), gene.name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting first optimization ==============\n",
      "1.385140400408024e-45 0.98\n",
      "Working with 2 zstates.\n",
      "0.00014247987205086657 0.98\n",
      "Working with 24 zstates.\n",
      "0.0025906558395850546 0.98\n",
      "Working with 135 zstates.\n",
      "0.024697271289687944 0.98\n",
      "Working with 194 zstates.\n",
      "0.03508719074260826 0.98\n",
      "Working with 196 zstates.\n",
      "0.03973851447855654 0.98\n",
      "Working with 197 zstates.\n",
      "0.04151498115030569 0.98\n",
      "Working with 197 zstates.\n",
      "0.042331079844077094 0.98\n",
      "Working with 197 zstates.\n",
      "0.04274874637125797 0.98\n",
      "Working with 197 zstates.\n",
      "0.043005388881112196 0.98\n",
      "Working with 197 zstates.\n",
      "0.043223925053593404 0.98\n",
      "Working with 197 zstates.\n",
      "0.04350822247904199 0.98\n",
      "Working with 197 zstates.\n",
      "0.0439728928144358 0.98\n",
      "Working with 197 zstates.\n"
     ]
    }
   ],
   "source": [
    "# perform the analysis\n",
    "\n",
    "print (\"Starting first optimization ==============\")\n",
    "emp_bayes = EmpiricalBayes(predictor, target, 1, init_params, method=\"new\")\n",
    "emp_bayes.fit()\n",
    "if zmax > 1:\n",
    "    if emp_bayes.success:\n",
    "        res = emp_bayes.params\n",
    "        print (\"Starting second optimization from previous results ================\")\n",
    "        # Python Error: C library could not compute z-components. Check C errors above.\n",
    "    else:\n",
    "        res = init_params\n",
    "        print (\"Starting second optimization from initial parameters ================\")\n",
    "    emp_bayes = EmpiricalBayes(predictor, target, zmax, res, method=\"new\")\n",
    "    emp_bayes.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PI: \t 0.000226820116643532\n",
      "mu: \t 0.0\n",
      "sigma: \t 0.001000044291115713\n",
      "sigmabg: \t 0.03400516423944316\n",
      "tau: \t 0.8749300969843092\n",
      "z-state:  0  Prob: 0.9760830572660332\n",
      "z-state:  1  Prob: 0.00022154478879425203\n",
      "z-state:  2  Prob: 0.0002214549039606261\n",
      "z-state:  3  Prob: 0.00022145491104533325\n",
      "z-state:  4  Prob: 0.00022145469584534133\n",
      "z-state:  5  Prob: 0.0002214546912404114\n",
      "z-state:  6  Prob: 0.00022145489721834484\n",
      "z-state:  7  Prob: 0.00022145099397045086\n",
      "z-state:  8  Prob: 0.00022145127034862408\n",
      "z-state:  9  Prob: 0.00022147731630414188\n",
      "z-state:  10  Prob: 0.00022146784510024799\n",
      "z-state:  11  Prob: 0.00022146808110090995\n",
      "z-state:  12  Prob: 0.00022144431983903975\n",
      "z-state:  13  Prob: 0.00022147789257922392\n",
      "z-state:  14  Prob: 0.00022144289727708054\n",
      "z-state:  15  Prob: 0.00022144279585390556\n",
      "z-state:  16  Prob: 0.00022144669761806142\n",
      "z-state:  17  Prob: 0.00022144243913084128\n",
      "z-state:  18  Prob: 0.00022145422694530245\n",
      "z-state:  19  Prob: 0.0002214550311682283\n",
      "z-state:  20  Prob: 0.00022145435138316974\n",
      "z-state:  21  Prob: 0.00022144330064260997\n",
      "z-state:  22  Prob: 0.00022144695735786636\n",
      "z-state:  23  Prob: 0.00022147290630593306\n",
      "z-state:  24  Prob: 0.0002214412399487801\n",
      "z-state:  25  Prob: 0.00022144203106603315\n",
      "z-state:  26  Prob: 0.0002214411138923316\n",
      "z-state:  27  Prob: 0.00022144109353990408\n",
      "z-state:  28  Prob: 0.00022145420136767413\n",
      "z-state:  29  Prob: 0.00022144226087147107\n",
      "z-state:  30  Prob: 0.00022144223521192052\n",
      "z-state:  31  Prob: 0.0002214422308602717\n",
      "z-state:  32  Prob: 0.00022144173612145314\n",
      "z-state:  33  Prob: 0.0002214419410841621\n",
      "z-state:  34  Prob: 0.00022144207783233316\n",
      "z-state:  35  Prob: 0.00022144207841200128\n",
      "z-state:  36  Prob: 0.00022144191916986252\n",
      "z-state:  37  Prob: 0.00022144193786999\n",
      "z-state:  38  Prob: 0.00022144118720179414\n",
      "z-state:  39  Prob: 0.00022144139997250353\n",
      "z-state:  40  Prob: 0.0002214443286989093\n",
      "z-state:  41  Prob: 0.00022144150414071107\n",
      "z-state:  42  Prob: 0.00022144267720097615\n",
      "z-state:  43  Prob: 0.0002214417846544609\n",
      "z-state:  44  Prob: 0.00022144163831328535\n",
      "z-state:  45  Prob: 0.0002214416125505643\n",
      "z-state:  46  Prob: 0.00022144113610672494\n",
      "z-state:  47  Prob: 0.00022144126519542154\n",
      "z-state:  48  Prob: 0.00022144110640200668\n",
      "z-state:  49  Prob: 0.00022146622738403385\n",
      "z-state:  50  Prob: 0.00022148026553627594\n",
      "z-state:  51  Prob: 0.00022144186121259224\n",
      "z-state:  52  Prob: 0.00022144201495415646\n",
      "z-state:  53  Prob: 0.00022144201495415646\n",
      "z-state:  54  Prob: 0.0002214524573908585\n",
      "z-state:  55  Prob: 0.00022147444665265962\n",
      "z-state:  56  Prob: 0.00022144526987313327\n",
      "z-state:  57  Prob: 0.00022144526987313327\n",
      "z-state:  58  Prob: 0.00022144526987313327\n",
      "z-state:  59  Prob: 0.00022144532670605516\n",
      "z-state:  60  Prob: 0.0002214451427112803\n",
      "z-state:  61  Prob: 0.0002214463323338262\n",
      "z-state:  62  Prob: 0.00022144712109171155\n",
      "z-state:  63  Prob: 0.0002214903938467639\n",
      "z-state:  64  Prob: 0.00022144365653418998\n",
      "z-state:  65  Prob: 0.00022146942605401392\n",
      "z-state:  66  Prob: 0.00022151863419180888\n",
      "z-state:  67  Prob: 0.00022144836458699163\n",
      "z-state:  68  Prob: 0.00022144130799498017\n",
      "z-state:  69  Prob: 0.00022144523672281978\n",
      "z-state:  70  Prob: 0.00022144380442351674\n",
      "z-state:  71  Prob: 0.00022153785629056756\n",
      "z-state:  72  Prob: 0.00022145824993987802\n",
      "z-state:  73  Prob: 0.00022147482408981558\n",
      "z-state:  74  Prob: 0.00022144950459924956\n",
      "z-state:  75  Prob: 0.00022144468383738467\n",
      "z-state:  76  Prob: 0.0002214471056126314\n",
      "z-state:  77  Prob: 0.0002214421675875206\n",
      "z-state:  78  Prob: 0.00022150108846294756\n",
      "z-state:  79  Prob: 0.00022144994494740814\n",
      "z-state:  80  Prob: 0.00022147276176259716\n",
      "z-state:  81  Prob: 0.00022144248607994832\n",
      "z-state:  82  Prob: 0.00022144243928748068\n",
      "z-state:  83  Prob: 0.0002214610272595763\n",
      "z-state:  84  Prob: 0.00022145283382943613\n",
      "z-state:  85  Prob: 0.0002214618162996714\n",
      "z-state:  86  Prob: 0.00022145707144535887\n",
      "z-state:  87  Prob: 0.00022145547688386837\n",
      "z-state:  88  Prob: 0.000221455225533293\n",
      "z-state:  89  Prob: 0.00022146115062712912\n",
      "z-state:  90  Prob: 0.00022144259820454224\n",
      "z-state:  91  Prob: 0.00022144232034054638\n",
      "z-state:  92  Prob: 0.00022144151698054508\n",
      "z-state:  93  Prob: 0.00022144113093483517\n",
      "z-state:  94  Prob: 0.00022144110206218708\n",
      "z-state:  95  Prob: 0.00022144139700656451\n",
      "z-state:  96  Prob: 0.0002214431988581501\n",
      "z-state:  97  Prob: 0.00022145807793001008\n",
      "z-state:  98  Prob: 0.00022145194219574507\n",
      "z-state:  99  Prob: 0.00022144449405932712\n",
      "z-state:  100  Prob: 0.00022145844382430818\n",
      "z-state:  101  Prob: 0.00022144545930077167\n",
      "z-state:  102  Prob: 0.00022145620668911372\n",
      "z-state:  103  Prob: 0.0002214577842534129\n",
      "z-state:  104  Prob: 0.00022144111014458866\n",
      "z-state:  105  Prob: 0.00022144111014458866\n",
      "z-state:  106  Prob: 0.00022144936126957104\n",
      "z-state:  107  Prob: 0.00022144984994416616\n",
      "z-state:  108  Prob: 0.00022144183866371376\n",
      "z-state:  109  Prob: 5.028317709540268e-08\n"
     ]
    }
   ],
   "source": [
    "if emp_bayes.success:\n",
    "    res = emp_bayes.params\n",
    "    res[4] = 1 / np.sqrt(res[4])\n",
    "    print(\"PI: \\t\",res[0])\n",
    "    print(\"mu: \\t\",res[1])\n",
    "    print(\"sigma: \\t\",res[2])\n",
    "    print(\"sigmabg: \\t\",res[3])\n",
    "    print(\"tau: \\t\",res[4])\n",
    "\n",
    "    model_snps = [snps[x] for x in snpmask]\n",
    "    model_zstates = list()\n",
    "    scaledparams = hyperparameters.scale(emp_bayes.params)\n",
    "    zprob, zexp = logmarglik.model_exp(scaledparams, predictor, target, emp_bayes.zstates)\n",
    "    for j, z in enumerate(emp_bayes.zstates):\n",
    "        this_zstate = ZstateInfo(state = z,\n",
    "                                 prob  = zprob[j],\n",
    "                                 exp   = list(zexp[j, :]) )\n",
    "        model_zstates.append(this_zstate)\n",
    "    # print(model_snps)\n",
    "    # for i,m in enumerate(model_zstates):\n",
    "    #     print(\"z-state: \",i,\" Prob:\", m.prob)\n",
    "    model.write_success_gene(gene, model_snps, model_zstates, res)\n",
    "else:\n",
    "    model.write_failed_gene(gene, np.zeros_like(init_params))\n",
    "    print (\"Failed optimization\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PI: \t 0.000226820116643532\n",
      "mu: \t 0.0\n",
      "sigma: \t 0.001000044291115713\n",
      "sigmabg: \t 0.03400516423944316\n",
      "tau: \t 0.8749300969843092\n"
     ]
    }
   ],
   "source": [
    "print(\"PI: \\t\",res[0])\n",
    "print(\"mu: \\t\",res[1])\n",
    "print(\"sigma: \\t\",res[2])\n",
    "print(\"sigmabg: \\t\",res[3])\n",
    "print(\"tau: \\t\",res[4])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "modelpath=\".\"\n",
    "cardio_home = os.path.join(home,\"datasets/cardiogenics/\")\n",
    "p_gtpath=os.path.join(cardio_home, \"genotypes/CG_\"+str(chrom)+\".imputed.gz\")\n",
    "p_samplepath=os.path.join(cardio_home, \"genotypes/CG.sample\")\n",
    "dataset=\"cardiogenics\"\n",
    "\n",
    "pred_path=os.path.join(home,\"gxpred/devtools\")\n",
    "outfileprefix = os.path.join(pred_path,\"pred_chr\"+str(chrom))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Actual Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Model\n",
    "from iotools.io_model import ReadModel\n",
    "from utils.containers import GeneExpressionArray\n",
    "from utils import mfunc\n",
    "\n",
    "p_model = ReadModel(modelpath, chrom)\n",
    "genes = p_model.genes\n",
    "gx = list()\n",
    "for gene in genes:\n",
    "\n",
    "    p_model.read_gene(gene)\n",
    "    p_model_snps = p_model.snps\n",
    "    p_model_zstates = p_model.zstates\n",
    "\n",
    "    x = gtutils.prediction_variables(p_snps, p_model_snps, p_genotype)\n",
    "    x = gtutils.normalize(p_model_snps, x)\n",
    "\n",
    "    ypred = np.zeros(p_nsample)\n",
    "    for z in p_model_zstates:\n",
    "        ypred += z.prob * np.dot(x.T, z.exp)\n",
    "\n",
    "    gx.append(GeneExpressionArray(geneid = gene.ensembl_id, expr_arr = ypred))\n",
    "\n",
    "\n",
    "# Write output\n",
    "mfunc.write_gcta_phenotype(outfileprefix, samplenames, gx)\n",
    "# printStamp(\"End Prediction\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparison to measured expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-02-22 18:42:43 - started reading genotype\n",
      "Read 1260727 snps in 786 samples.\n",
      "2018-02-22 18:55:32 - Finished readings snps\n"
     ]
    }
   ],
   "source": [
    "# read Cardiogenics Genotype\n",
    "\n",
    "p_oxf = ReadOxford(p_gtpath, p_samplepath, chrom, dataset)\n",
    "p_genotype = np.array(p_oxf.dosage)\n",
    "p_samplenames = p_oxf.samplenames\n",
    "p_snps = p_oxf.snps_info\n",
    "p_nsample = len(p_oxf.samplenames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Cardiogenics expression data\n",
    "\n",
    "from iotools.readrpkm import ReadRPKM\n",
    "expdatapath = os.path.join(pred_home, \"expression/lumi_mono_all_vsn_batches_adjusted.RData.exp.txt\")\n",
    "\n",
    "# Load dataset Gene Expression\n",
    "rpkm = ReadRPKM(expdatapath, \"cardiogenics\")\n",
    "expression = rpkm.expression\n",
    "expr_donors = rpkm.donor_ids\n",
    "gene_names = rpkm.gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "class ReadPrediction:\n",
    "    \n",
    "    \n",
    "    def __init__(self, predpath, samplepath, predictor):\n",
    "        self._pred_path = predpath\n",
    "        self._predictor = predictor #gxpred or predixcan\n",
    "        self._expr_mat = None\n",
    "        self._samplepath = samplepath\n",
    "        self._load_samples()\n",
    "        self._load_chromosomes()\n",
    "    \n",
    "    def _load_samples(self):\n",
    "        samples = list()\n",
    "        with open(self._samplepath, 'r') as instream:\n",
    "            for line in instream:\n",
    "                if re.search('^#', line):\n",
    "                    continue\n",
    "                arr = line.strip().split()\n",
    "                samples.append(arr[0])\n",
    "        self._samples = samples            \n",
    "    \n",
    "    # return expression matrix in samples-by-genes format\n",
    "    def _load_chromosome(self, chrom):\n",
    "        \n",
    "        if self._predictor == \"gxpred\":\n",
    "            basefile = \"pred_chr\"+chrom\n",
    "            pred_geneids = os.path.join(self._pred_path,basefile+\".geneids\")\n",
    "            pred_exp = os.path.join(self._pred_path,basefile+\".txt\")\n",
    "            \n",
    "            with open(pred_geneids, 'r') as infile:\n",
    "                gene_names = [l.strip() for l in infile.readlines()]\n",
    "                ncolumns = len(gene_names)\n",
    "            \n",
    "            # need the number of columns (num of genes), to skip first two sample id columns\n",
    "            expr_mat = np.loadtxt(pred_exp, usecols=range(2,ncolumns+2))            \n",
    "            \n",
    "        if self._predictor == \"predixcan\":\n",
    "            filename = \"predixcan_chr\"+chrom+\".output\"\n",
    "            filepath = os.path.join(self._pred_path, filename)\n",
    "            with open(filepath, 'r') as instream:\n",
    "                gene_names = instream.readline().strip().split()\n",
    "            expr_mat = np.loadtxt(filepath, skiprows=1)\n",
    "            \n",
    "        print(expr_mat.shape)\n",
    "        print(\"genes: \",len(gene_names))\n",
    "        return expr_mat, gene_names\n",
    "                \n",
    "    def _load_chromosomes(self):\n",
    "        self._gene_names = []\n",
    "        self._expr_mat = None\n",
    "               \n",
    "        for chrom in range(1,23):\n",
    "            expr_mat, gene_names = self._load_chromosome(str(chrom))\n",
    "            if self._predictor == \"gxpred\":\n",
    "                if type(self._expr_mat) is not np.ndarray:\n",
    "                    self._expr_mat = expr_mat.T\n",
    "                else:\n",
    "                    self._expr_mat = np.concatenate((self._expr_mat, expr_mat.T), axis=0)\n",
    "                self._gene_names += gene_names\n",
    "            \n",
    "            if self._predictor == \"predixcan\":\n",
    "                if type(self._expr_mat) is not np.ndarray:\n",
    "                    self._expr_mat = expr_mat\n",
    "                else:\n",
    "                    self._expr_mat += expr_mat\n",
    "                self._gene_names = gene_names\n",
    "\n",
    "    def sort_by_samples(self, samplelist, use_prev = False):\n",
    "        # sorts expression matrix by the samplelist\n",
    "        # writes to sorted_expr_mat\n",
    "        # if use_prev == True, should use previously sorted matrix, samples and gene\n",
    "        # else: use original expr_mat, genes and samples\n",
    "        \n",
    "        # sort sample names for px gx and cardiogenics\n",
    "\n",
    "        # since they are the same, the index will be same for both (pgx)\n",
    "\n",
    "        common_samples = [x for x in expr_donors if x in gx_samples]\n",
    "        rgx_samples_index = [gx_samples.index(i) for i in common_samples]\n",
    "        cardio_samples_index = [expr_donors.index(i) for i in common_samples]\n",
    "\n",
    "        rgx_sorted_samples = [gx_samples[i] for i in rgx_samples_index]\n",
    "        cardio_sorted_samples = [expr_donors[i] for i in cardio_samples_index]\n",
    "\n",
    "        # are Samples indexes ok?\n",
    "        print(rgx_sorted_samples == cardio_sorted_samples)\n",
    "        print(len(rgx_sorted_samples))\n",
    "        print(len(cardio_sorted_samples))\n",
    "        pass\n",
    "    \n",
    "    def sort_by_gene(self, genelist, use_prev = False):\n",
    "        # sorts expression matrix by the given genelist\n",
    "        \n",
    "        # sort gene names\n",
    "        print(\"Total cardio genes:\",len(gene_names))\n",
    "\n",
    "        gx_gene_names2 = [x.split(\".\")[0] for x in gx_gene_batch]\n",
    "        rand_gx_gene_names2 = [x.split(\".\")[0] for x in rand_gx_gene_batch]\n",
    "\n",
    "        # I don't know WHY random gxpred has less genes!! intersection should be 11642 and not 11179\n",
    "        #set_a = set(gx_gene_batch)\n",
    "        #set_b = set(rand_gx_gene_batch)\n",
    "        #set_c = set_a.intersection(set_b)\n",
    "        #print(len(set_a), len(set_b), len(set_c))\n",
    "\n",
    "        set_rgx = set(rand_gx_gene_names2)\n",
    "        set_gx = set(gx_gene_names2)\n",
    "\n",
    "        common_genes = set_rgx.intersection(set_gx)\n",
    "        print(\"rgx and gx common genes: \",len(common_genes))\n",
    "\n",
    "        cardio_common_genes = [g for g in common_genes if g in gene_names]\n",
    "\n",
    "        print(\"cardio,rgx,gx, common genes: \", len(cardio_common_genes))\n",
    "\n",
    "        # this is for indexing cardiogenics expression matrix\n",
    "        gx_cardio_gene_index = [gene_names.index(i) for i in cardio_common_genes]\n",
    "        print(len(gx_cardio_gene_index),\"/\",len(gene_names))\n",
    "\n",
    "        # this is for indexing gx and px gene names and expression mat\n",
    "        gx_gene_index = [gx_gene_names2.index(i) for i in cardio_common_genes]\n",
    "        rgx_gene_index = [rand_gx_gene_names2.index(i) for i in cardio_common_genes]\n",
    "\n",
    "        print(len(gx_gene_index),\"/\",len(gx_gene_batch))\n",
    "        print(len(rgx_gene_index),\"/\",len(rand_gx_gene_batch))\n",
    "        pass\n",
    "    \n",
    "    \n",
    "    \n",
    "    @property\n",
    "    def expr_mat(self):\n",
    "        return self._expr_mat\n",
    "    \n",
    "    @property\n",
    "    def gene_names(self):\n",
    "        return self._gene_names\n",
    "    \n",
    "    @property\n",
    "    def samples(self):\n",
    "        return self._samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(786, 1554)\n",
      "genes:  1554\n",
      "(786, 952)\n",
      "genes:  952\n",
      "(786, 835)\n",
      "genes:  835\n",
      "(786, 523)\n",
      "genes:  523\n",
      "(786, 639)\n",
      "genes:  639\n",
      "(786, 709)\n",
      "genes:  709\n",
      "(786, 614)\n",
      "genes:  614\n",
      "(786, 502)\n",
      "genes:  502\n",
      "(786, 592)\n",
      "genes:  592\n",
      "(786, 556)\n",
      "genes:  556\n",
      "(786, 824)\n",
      "genes:  824\n",
      "(786, 830)\n",
      "genes:  830\n",
      "(786, 245)\n",
      "genes:  245\n",
      "(786, 487)\n",
      "genes:  487\n",
      "(786, 445)\n",
      "genes:  445\n",
      "(786, 672)\n",
      "genes:  672\n",
      "(786, 864)\n",
      "genes:  864\n",
      "(786, 203)\n",
      "genes:  203\n",
      "(786, 1160)\n",
      "genes:  1160\n",
      "(786, 395)\n",
      "genes:  395\n",
      "(786, 148)\n",
      "genes:  148\n",
      "(786, 375)\n",
      "genes:  375\n"
     ]
    }
   ],
   "source": [
    "gxpred_predpath = \"/home/franco/cbscratch/datasets/cardiogenics/gxpred_predictions_klinikum\"\n",
    "samplepath = \"/home/franco/cluster2/datasets/cardiogenics/genotypes/CG.sample\"\n",
    "gxpred = ReadPrediction(gxpred_predpath, samplepath, \"gxpred\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14124, 786)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gxpred.expr_mat.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n",
      "(786, 6759)\n",
      "genes:  6759\n"
     ]
    }
   ],
   "source": [
    "pxpred_predpath = \"/home/franco/cbscratch/datasets/cardiogenics/predixcan_predictions_klinikum\"\n",
    "samplepath = \"/home/franco/cluster2/datasets/cardiogenics/genotypes/CG.sample\"\n",
    "predixcanpred = ReadPrediction(pxpred_predpath, samplepath, \"predixcan\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(786, 6759)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predixcanpred.expr_mat.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[GeneExpressionArray(geneid='ENSG00000183386.5', expr_arr=array([ 5.76509206e-02,  2.30666594e-01, -1.03668846e-01, -3.38784261e-01,\n",
       "        1.68540142e-01, -2.97173274e-03, -2.17672991e-01, -1.81710250e-01,\n",
       "       -1.38896980e-01, -1.05242816e-01,  6.79431197e-02, -1.29499674e-01,\n",
       "       -2.76376253e-01, -1.44089024e-01,  1.64167333e-01, -3.34455286e-01,\n",
       "        1.33228357e-01, -2.80445636e-01, -2.09993032e-01, -2.17784174e-01,\n",
       "       -4.25467932e-02,  7.11238411e-01,  5.99397386e-03, -5.13565772e-02,\n",
       "       -3.27922275e-01,  1.67855406e-01, -2.41047709e-01,  2.72924269e-01,\n",
       "        9.14703384e-02, -2.73657620e-01,  2.63657323e-01, -4.24768146e-01,\n",
       "        3.64469610e-02,  8.68841089e-02, -3.00838751e-02, -1.35982421e-01,\n",
       "        3.70550094e-03, -2.03859379e-01, -2.19058561e-01,  2.09547526e-01,\n",
       "       -8.56014635e-02,  2.67564629e-01, -1.54836684e-01,  6.21477823e-02,\n",
       "        2.48127600e-01,  2.45707722e-02,  2.08148523e-01, -2.28517779e-01,\n",
       "       -1.85447030e-01,  3.77538157e-01, -8.47015289e-02,  2.45666588e-01,\n",
       "        5.76185809e-02,  8.11065087e-03, -2.79581439e-01,  9.32478415e-02,\n",
       "        2.31784968e-01, -1.11786431e-01, -7.68768184e-02,  9.92213097e-03,\n",
       "       -2.24126258e-01,  3.97198152e-01,  2.71130193e-02, -4.50726979e-01,\n",
       "       -1.39302707e-01, -1.72037045e-01, -2.65115329e-01, -1.80341946e-01,\n",
       "        1.00148006e-01, -1.01015377e-01, -1.30939295e-01, -3.01140839e-01,\n",
       "       -2.91040464e-01,  2.43916669e-01, -3.87079781e-02, -3.00268390e-01,\n",
       "        3.20037761e-02,  1.26293361e-01, -6.10536130e-02, -2.05381833e-01,\n",
       "       -1.34444415e-01, -1.94838724e-01, -3.96369845e-01, -1.06867171e-01,\n",
       "        2.07028148e-01,  1.19403143e-01,  1.54509775e-01, -2.03488927e-01,\n",
       "        4.71379575e-02, -8.85881509e-02, -1.17793713e-02, -2.16073418e-01,\n",
       "        4.59023041e-01, -4.79155544e-02,  1.13788100e-02,  1.03651958e-01,\n",
       "        7.95052453e-02,  6.39006827e-02, -7.59670428e-02, -1.66246855e-01,\n",
       "        2.32668938e-02, -3.40545811e-02, -5.88228536e-02, -1.27989049e-01,\n",
       "        6.02607464e-02,  1.21375499e-01,  1.83296434e-01, -3.63273149e-01,\n",
       "       -3.19786995e-01, -1.04656970e-01,  9.52708794e-02, -1.44250280e-02,\n",
       "       -8.67170494e-02,  3.37639054e-01,  2.91017637e-01, -9.93523983e-02,\n",
       "        1.13637534e-01,  8.42609765e-02, -1.16775901e-01,  1.80645883e-01,\n",
       "       -1.79661181e-01, -1.05988235e-01, -2.14398526e-02,  2.09801058e-01,\n",
       "       -2.19684170e-01, -2.05895930e-01,  5.09030599e-02,  2.22169280e-02,\n",
       "        4.94421888e-02, -3.68899837e-02, -1.29378752e-01, -3.66822400e-01,\n",
       "        1.84675262e-01,  1.10560357e-01,  2.69367964e-01, -3.13319989e-01,\n",
       "       -5.39665727e-02,  9.70700454e-02, -2.65721383e-01, -1.29858877e-01,\n",
       "       -7.53647894e-02,  1.43969654e-01, -5.15772426e-01, -1.42036052e-01,\n",
       "       -1.80365337e-01,  8.93267695e-03, -1.57943131e-02, -1.62490041e-01,\n",
       "       -1.20346508e-01, -3.17885716e-04,  1.13656616e-01,  1.24382509e-01,\n",
       "        1.63499601e-01,  8.16454644e-02, -1.13492751e-01,  1.42360721e-01,\n",
       "        4.22820696e-02,  5.37400341e-02,  1.73823653e-01,  1.56224372e-01,\n",
       "       -1.50690565e-01, -1.67688146e-01,  3.04321104e-01, -7.73976009e-02,\n",
       "       -2.32333121e-02, -1.24673265e-02, -7.88631414e-02,  3.07983488e-03,\n",
       "        1.65129634e-02, -1.79109982e-01, -1.13625110e-01, -4.25745042e-02,\n",
       "        1.14244761e-01,  3.70686641e-02, -2.70805701e-01, -2.24214967e-01,\n",
       "       -3.86815440e-01, -4.80516657e-02, -1.32939116e-01,  2.57954220e-01,\n",
       "       -9.83531566e-03, -7.65410899e-02, -3.95469763e-01,  1.35137386e-01,\n",
       "        1.38423040e-01, -1.52832669e-01,  6.90148143e-02,  2.80994600e-01,\n",
       "       -2.91102474e-02, -3.36502365e-01, -3.38859971e-01,  9.07547921e-02,\n",
       "       -2.80797899e-02,  8.26458466e-03,  1.83082659e-01, -8.69363934e-02,\n",
       "        6.31412835e-02, -2.98922713e-01,  5.37297907e-01, -1.69879309e-01,\n",
       "       -1.90170536e-01, -6.29140358e-02, -8.79597231e-02, -1.01902690e-01,\n",
       "        1.46017074e-01,  1.61636940e-01,  6.07646017e-02, -9.52503849e-03,\n",
       "        1.53571941e-01,  1.32829634e-01, -7.90495940e-02, -3.03730416e-01,\n",
       "       -4.26744561e-02,  3.32675746e-01,  7.09315734e-02,  2.26056114e-01,\n",
       "        1.07438307e-01,  1.34584648e-02, -1.17807905e-01, -6.18344041e-02,\n",
       "        6.84208325e-02,  1.94768749e-01,  3.29990865e-01,  3.42470644e-02,\n",
       "       -1.15437032e-01, -2.28364023e-01, -3.76916926e-02, -2.69583233e-01,\n",
       "        1.47378985e-01, -3.37623712e-01, -1.59458762e-02,  8.05170094e-02,\n",
       "       -2.32563375e-01, -3.81995294e-02,  2.89759060e-01, -1.18775602e-01,\n",
       "       -2.59731781e-01,  1.74869557e-01, -4.82904823e-02,  1.32666488e-01,\n",
       "        4.69920327e-02, -5.62494867e-02, -6.37173011e-02,  2.46821424e-01,\n",
       "        5.78663300e-02, -2.43163478e-01,  1.92210147e-01,  9.93520423e-03,\n",
       "        8.92053078e-02,  7.21242297e-02, -1.59025646e-01, -3.49269410e-02,\n",
       "        1.21780205e-01, -4.91774528e-02,  3.68203714e-01, -3.83807108e-02,\n",
       "        2.90702929e-01,  3.22104819e-01,  1.47601279e-01,  3.38336604e-01,\n",
       "        1.75423508e-01,  8.56463337e-02, -2.44363313e-01,  6.95371906e-01,\n",
       "       -3.85892787e-02,  4.47313761e-02,  7.97541516e-02,  3.97121379e-01,\n",
       "        9.38101680e-02, -3.68103060e-02, -1.88493131e-01,  5.08806688e-02,\n",
       "       -6.27735152e-02, -2.98896160e-01, -1.59899864e-01, -2.35946420e-02,\n",
       "       -4.01752389e-03, -1.62004675e-01, -2.11305972e-01, -2.57743848e-02,\n",
       "        3.58419410e-01, -9.26010148e-02, -6.91720526e-02, -1.51841059e-01,\n",
       "        1.16194618e-01, -9.44870637e-02, -5.43680001e-02,  3.48122562e-01,\n",
       "        7.83044711e-02,  9.78206424e-04,  1.69414938e-01,  1.44107065e-01,\n",
       "        7.56535498e-02,  2.56274980e-01,  1.26544970e-01,  9.11803513e-02,\n",
       "       -2.27676913e-01,  1.04935873e-01,  1.44045752e-02,  7.56263734e-02,\n",
       "        2.81980373e-01,  3.06274514e-01, -2.37321062e-01, -8.00251969e-03,\n",
       "        8.83535695e-02, -2.34481827e-01,  1.65170816e-01, -1.22011783e-01,\n",
       "       -2.84562252e-02, -3.26515140e-02, -3.54033734e-02, -8.03690515e-02,\n",
       "       -1.05778962e-01,  1.79585092e-01, -6.23776600e-02,  2.38519197e-02,\n",
       "       -2.20030899e-01, -6.50916353e-02, -1.55435417e-01,  2.53138271e-01,\n",
       "       -3.91918342e-02, -2.53567713e-01, -6.72935861e-02, -1.31366875e-01,\n",
       "       -2.50703220e-01, -1.05309796e-01, -1.06209228e-01, -1.93797954e-01,\n",
       "        2.37429213e-02,  4.62379339e-02,  1.55238979e-02, -7.08015244e-03,\n",
       "        1.23605167e-01,  6.33242770e-02, -2.72234365e-01, -2.60727295e-01,\n",
       "       -3.51868549e-01,  3.96680386e-02, -2.27536133e-01, -1.00439627e-01,\n",
       "       -1.24488330e-02,  2.30998611e-01, -3.70962715e-03,  1.26721039e-01,\n",
       "        1.36569241e-01,  2.04620064e-01,  4.01404000e-02,  4.45651313e-01,\n",
       "        1.19278689e-01,  1.61996501e-01, -1.84851982e-01,  2.62038304e-01,\n",
       "        2.52849359e-01,  1.86122025e-01, -4.23810691e-02, -7.85253311e-02,\n",
       "       -3.99223217e-01,  9.53797057e-03, -2.06143354e-01,  4.57559550e-02,\n",
       "       -9.46326001e-02,  3.10741505e-01,  1.69803964e-01, -1.80266481e-01,\n",
       "        3.11575923e-01,  1.48151207e-01,  6.48966746e-02, -9.76362524e-02,\n",
       "       -1.91628857e-01, -6.09778193e-02,  6.52825553e-02, -3.12966246e-01,\n",
       "       -1.13719404e-01,  2.10542392e-01,  5.38830878e-02,  1.16798287e-02,\n",
       "       -9.59859896e-02, -9.03719048e-02, -1.46053635e-01,  8.79741759e-02,\n",
       "        1.13583398e-01, -1.54308333e-01, -1.68085874e-01,  2.36262156e-01,\n",
       "       -7.67566079e-02, -1.42049257e-01,  2.15734661e-02,  1.76616355e-01,\n",
       "        6.36892616e-02, -4.20849537e-02, -1.61098699e-01,  2.08596581e-01,\n",
       "       -2.08815208e-01, -1.60642193e-01,  2.73907914e-02,  4.09601542e-02,\n",
       "       -6.76924384e-02, -1.33063369e-01, -1.39835055e-01, -2.96817072e-01,\n",
       "       -7.28725145e-02,  1.80777777e-01, -2.89658968e-01, -4.69932067e-02,\n",
       "       -2.07753813e-01, -1.59873179e-01, -1.64454831e-01,  1.18466897e-02,\n",
       "       -5.69980709e-02, -2.25600258e-02, -2.01853148e-01, -2.95330128e-01,\n",
       "       -2.04157219e-01,  2.14708270e-01, -2.80524571e-01,  3.66754377e-03,\n",
       "       -6.94662642e-02,  6.17154775e-03, -1.73035801e-01, -1.52168630e-01,\n",
       "        3.11972310e-01,  1.61084178e-01,  1.07634424e-01,  1.23281907e-01,\n",
       "       -2.77445494e-01, -3.41909777e-02, -3.10384521e-01,  2.23392138e-01,\n",
       "       -1.49661714e-01,  2.95284782e-01, -1.79418969e-01,  2.11832650e-01,\n",
       "       -4.19380802e-01,  2.03084311e-02,  1.75858245e-01,  1.97276512e-01,\n",
       "        1.39812656e-01,  1.07338889e-01, -8.93980284e-02, -4.32282474e-02,\n",
       "       -1.18382955e-01,  6.53279290e-03,  7.34219808e-02, -3.48206698e-01,\n",
       "       -4.38110273e-02,  1.32413465e-01,  8.27939971e-02,  1.60686294e-01,\n",
       "       -3.09327409e-01,  5.83740784e-02,  1.48854673e-01,  9.59447496e-02,\n",
       "       -9.87712142e-02, -8.27355462e-02,  8.89019396e-02,  1.48745703e-02,\n",
       "        6.34144639e-02,  2.02419125e-02,  4.38251937e-02,  3.39803836e-03,\n",
       "        7.12154883e-02, -5.26718415e-01,  7.23137221e-02, -2.59194372e-01,\n",
       "       -1.24959645e-01,  1.06071680e-01, -1.04049882e-01, -3.34112785e-01,\n",
       "       -3.65766610e-02,  1.59591332e-01, -9.70430319e-02, -4.20616888e-01,\n",
       "       -1.34373683e-01,  1.70372897e-01,  4.73118225e-02, -1.37199157e-01,\n",
       "        2.77782923e-02,  7.91019031e-02, -2.95181853e-02,  1.86646535e-01,\n",
       "       -1.01958975e-01,  1.12415538e-01,  8.61255523e-03, -2.56308343e-01,\n",
       "       -5.31384609e-01,  2.11524896e-01, -1.92966423e-01,  2.19724071e-01,\n",
       "       -9.12868856e-02, -2.69618167e-01, -8.12803593e-02, -2.66166911e-01,\n",
       "        1.14207308e-01, -2.52550635e-01,  2.17739590e-01,  2.82606449e-01,\n",
       "        1.47620150e-01,  1.83569830e-01,  1.55637580e-01, -3.05079888e-02,\n",
       "       -2.05907730e-01, -7.26024045e-02, -1.18393842e-01,  8.04947552e-02,\n",
       "        1.81770126e-01, -7.14093722e-03,  1.96055547e-01, -1.26049446e-01,\n",
       "        1.52005273e-01,  2.29754635e-02, -4.18152218e-02,  4.34568315e-03,\n",
       "        2.80720230e-01, -1.77536290e-02,  5.66906654e-02,  1.23584837e-01,\n",
       "        1.89559576e-02,  4.18996512e-02, -5.70967540e-02,  2.08647177e-01,\n",
       "        3.65820843e-01, -1.50193308e-01,  1.17894548e-01, -1.27082324e-01,\n",
       "       -8.54416269e-02, -2.16820442e-02,  2.27320247e-01, -1.35725105e-01,\n",
       "       -2.77816523e-02,  5.47066259e-02, -4.65300785e-01, -8.85449069e-02,\n",
       "       -2.98208881e-02,  1.34916722e-02,  2.04317639e-01,  1.91594994e-01,\n",
       "       -9.82755931e-02, -6.21526106e-02, -3.11060101e-01, -4.46274851e-02,\n",
       "        1.09127363e-01, -2.09058953e-01, -1.09725840e-01, -1.74901748e-02,\n",
       "       -1.45630825e-01, -1.03693817e-01, -1.38830392e-01, -8.72509357e-02,\n",
       "        8.63263828e-02, -3.14854284e-01,  2.21838724e-01, -2.12893965e-02,\n",
       "        2.23448145e-01, -1.08799559e-01, -2.69842778e-01,  3.86478841e-02,\n",
       "        2.00608849e-02, -1.19348616e-01, -2.64599799e-01, -1.94129837e-01,\n",
       "       -2.84699413e-02,  2.91500798e-02, -6.27587134e-02,  2.44171071e-01,\n",
       "       -2.04563588e-02,  8.89641900e-02,  6.25124166e-02, -1.43751468e-01,\n",
       "       -6.27485807e-02,  6.63264884e-02,  1.14456680e-01, -1.36097680e-02,\n",
       "        2.61876089e-01,  1.16895338e-01, -7.28373573e-02,  7.00700188e-03,\n",
       "       -1.38763339e-02, -1.03055666e-01, -1.43272062e-02,  9.95540541e-02,\n",
       "       -1.90276591e-01,  1.65043377e-01, -3.62733344e-03, -4.99893382e-02,\n",
       "       -4.82383122e-01, -4.05643034e-01,  1.36428503e-01, -2.18551884e-02,\n",
       "       -4.10707276e-02, -2.51451078e-01, -3.54809759e-02,  1.29846519e-01,\n",
       "        2.45414085e-01, -1.61470011e-01,  2.58504036e-01,  2.90374890e-01,\n",
       "       -9.17902500e-02,  1.39969314e-01, -2.83660706e-02, -2.30955611e-01,\n",
       "        4.37572757e-03,  8.03966194e-02, -1.21598145e-01,  2.00575861e-01,\n",
       "       -1.58033985e-01,  3.81871605e-01, -5.98682928e-02,  3.30228926e-01,\n",
       "       -8.13480324e-02, -7.77350559e-02,  1.89975441e-03,  2.22957161e-01,\n",
       "       -1.91943691e-02,  1.05658401e-01, -1.47064097e-01,  1.61882401e-02,\n",
       "       -1.61867693e-01,  4.59680977e-02, -2.64475470e-01, -1.37563223e-01,\n",
       "        7.73577710e-02, -9.77844577e-02,  3.10380910e-01,  1.30045690e-03,\n",
       "        1.40590804e-01, -1.39860672e-01, -1.43309232e-01, -2.92704416e-01,\n",
       "       -3.61114661e-01, -2.18233179e-01,  6.59160449e-02, -1.54487307e-02,\n",
       "        1.77825728e-01,  6.91944550e-03,  6.76666558e-02,  9.86803679e-03,\n",
       "        7.84862785e-02,  8.97648450e-02,  6.38271971e-02, -7.62620174e-02,\n",
       "        7.03168510e-02, -7.96143422e-02,  8.87767885e-02,  1.16001395e-01,\n",
       "        2.31749390e-01,  1.71049919e-01,  2.38326569e-01,  9.47607173e-02,\n",
       "        1.90467658e-01,  3.33614281e-02, -3.03836476e-02, -3.67304362e-01,\n",
       "       -1.78448955e-01, -1.16076711e-01, -7.78414895e-02,  8.33659445e-02,\n",
       "        2.00460248e-01, -6.25191611e-02,  1.01506401e-01, -2.42817308e-02,\n",
       "        1.89563738e-01,  2.33849785e-01, -3.85703325e-02,  2.66472349e-01,\n",
       "        1.24408562e-01, -8.17706281e-02,  2.63558668e-01, -2.93613673e-01,\n",
       "        5.37394944e-02, -1.70863362e-01, -1.37236681e-01,  6.19527146e-01,\n",
       "        4.86797522e-03, -2.72657056e-01,  5.74248076e-02,  8.37015510e-02,\n",
       "       -2.19156447e-02,  1.96630410e-01,  4.76529869e-02, -1.91318611e-01,\n",
       "        3.21305261e-02,  2.71468639e-01,  2.92640351e-01,  1.67429800e-01,\n",
       "       -5.65147682e-02,  3.16653636e-02, -1.83028318e-01,  2.77802563e-01,\n",
       "       -7.65267360e-02, -1.74288572e-01,  9.66813037e-02,  2.66928069e-01,\n",
       "        2.25388671e-01, -1.34458361e-01,  1.27854844e-01,  1.89233773e-02,\n",
       "       -6.91270898e-02,  7.96305829e-02,  2.43098222e-01,  1.10502194e-01,\n",
       "        1.42260520e-01,  8.77762865e-02,  3.05768394e-03,  5.84530321e-02,\n",
       "        1.10916144e-01, -2.19333115e-01, -1.50408947e-01, -7.43870309e-02,\n",
       "        6.56997174e-02,  1.47658221e-01,  1.42928107e-01,  2.98591231e-02,\n",
       "       -2.51691591e-02, -4.14933412e-01,  1.08167484e-01, -6.17819654e-02,\n",
       "       -2.14366019e-01, -3.45836088e-02,  3.10431964e-02, -1.58729201e-01,\n",
       "       -1.88010966e-01, -2.12074682e-01, -2.06324219e-01,  1.62596664e-01,\n",
       "        1.00184549e-01, -7.73462514e-02,  2.15897800e-01, -2.14877061e-01,\n",
       "        1.71434142e-02,  3.26451468e-02,  7.45014025e-02, -7.96976506e-02,\n",
       "        4.67902121e-03, -5.11497250e-02,  1.02253767e-02,  9.28408768e-02,\n",
       "        2.56571635e-01, -4.31170131e-01, -8.00735485e-02,  2.09239471e-01,\n",
       "        1.42594721e-01, -1.63485596e-01,  1.88795240e-02, -9.55115320e-02,\n",
       "        3.15355573e-02,  2.46197183e-02, -1.24220168e-01, -1.65098679e-01,\n",
       "       -2.33016743e-02,  4.26517735e-02, -3.84969721e-02, -8.16696221e-02,\n",
       "       -1.00261235e-01,  3.35812611e-02, -1.88530316e-01, -1.80443430e-01,\n",
       "       -2.88887050e-01,  3.67105681e-02,  1.23576234e-01,  9.07463433e-02,\n",
       "       -4.10706355e-02,  1.21922012e-01, -4.24897785e-01, -6.55528392e-02,\n",
       "        1.05907597e-01,  9.55123824e-02, -4.43838798e-01, -6.27786419e-02,\n",
       "        2.94664944e-01, -1.03694948e-01, -9.36531665e-02, -5.05566671e-02,\n",
       "       -3.78403927e-01,  1.11162953e-02, -4.91432747e-03,  1.48166538e-01,\n",
       "        4.58187966e-03, -3.69888048e-02, -2.37606925e-02, -2.89801041e-01,\n",
       "        2.21869555e-01, -2.19184006e-01, -1.22803676e-01,  6.03787931e-02,\n",
       "       -4.60884001e-01, -1.51891416e-01]))]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import matplotlib.mlab as mlab\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "ax.scatter(, gx.expr_arr)\n",
    "ax.plot([0,1], [0,1])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
