{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "\n",
    "\n",
    "import sys, os\n",
    "sys.path.append(\"../\")\n",
    "\n",
    "import numpy as np\n",
    "import math\n",
    "import pickle\n",
    "from config_annots import *\n",
    "from iotools.readOxford import ReadOxford\n",
    "from iotools.readrpkm import ReadRPKM\n",
    "from iotools.io_model import WriteModel\n",
    "from inference.linreg_association import LinRegAssociation\n",
    "from inference.empirical_bayes import EmpiricalBayes\n",
    "from utils import hyperparameters\n",
    "from inference import logmarglik\n",
    "from iotools import readgtf\n",
    "from utils import gtutils\n",
    "from utils import mfunc\n",
    "from utils.containers import ZstateInfo\n",
    "from utils.printstamp import printStamp\n",
    "from helper_functions import load_target_genes, write_params\n",
    "from sklearn.preprocessing import scale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation (use complete gene name in gtf without trimming the version)\n",
    "# load annotation for whole genome\n",
    "gene_info = readgtf.gencode_v12(gtfpath, trim=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 70 genes with high r2 values\n",
      "\n",
      "Found 7 genes in CHR 12\n",
      "['ENSG00000171860.4', 'ENSG00000064115.6', 'ENSG00000139610.1', 'ENSG00000090382.2', 'ENSG00000127337.2', 'ENSG00000135643.4', 'ENSG00000184752.8']\n"
     ]
    }
   ],
   "source": [
    "# Load gene list\n",
    "# genelistfile = \"genes4testing_high_and_low_r2_0.001\"\n",
    "genelistfile = \"genes4testing_highr2\"\n",
    "selected_gene_ids = load_target_genes(genelistfile, gene_info, chrom)\n",
    "print(selected_gene_ids)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-03-14 20:31:07 - Reading pickled genotype\n",
      "2018-03-14 20:31:10 - Done reading\n",
      "2018-03-14 20:31:17 - Selection of samples\n"
     ]
    }
   ],
   "source": [
    "\n",
    "if not os.path.exists(learn_pickfile):\n",
    "    # read Genotype\n",
    "    oxf = ReadOxford(gtex_gtpath, gtex_samplepath, chrom, learning_dataset)\n",
    "    genotype = np.array(oxf.dosage)\n",
    "    samplenames = oxf.samplenames\n",
    "    snps = oxf.snps_info\n",
    "\n",
    "    printStamp(\"Dumping CHR {:d} genotype\".format(chrom))\n",
    "    with open(learn_pickfile, 'wb') as output:\n",
    "        pickle.dump(oxf, output, pickle.HIGHEST_PROTOCOL)\n",
    "else:\n",
    "    printStamp(\"Reading pickled genotype\")\n",
    "    with open(learn_pickfile, 'rb') as input:\n",
    "        pickled_oxf = pickle.load(input)\n",
    "\n",
    "    printStamp(\"Done reading\")\n",
    "\n",
    "    genotype = np.array(pickled_oxf.dosage)\n",
    "    samplenames = pickled_oxf.samplenames\n",
    "    snps = pickled_oxf.snps_info\n",
    "    nsample = len(pickled_oxf.samplenames)\n",
    "\n",
    "# Quality control\n",
    "f_snps, f_genotype = gtutils.remove_low_maf(snps, genotype, 0.1)\n",
    "gt = gtutils.normalize(f_snps, f_genotype)\n",
    "\n",
    "# Gene Expression\n",
    "rpkm = ReadRPKM(gtex_rpkmpath, \"gtex\")\n",
    "expression = rpkm.expression\n",
    "expr_donors = rpkm.donor_ids\n",
    "gene_names = rpkm.gene_names\n",
    "\n",
    "# Selection\n",
    "printStamp(\"Selection of samples\")\n",
    "vcfmask, exprmask = mfunc.select_donors(samplenames, expr_donors)\n",
    "genes, indices = mfunc.select_genes(gene_info, gene_names)\n",
    "\n",
    "gene_training_list = []\n",
    "for i, gene in enumerate(genes):\n",
    "    k = indices[i]\n",
    "    if gene.ensembl_id in selected_gene_ids and gene.chrom == chrom:\n",
    "        gene_training_list.append((k,gene))\n",
    "        # print(k,gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters[0] = ['gxpred-bslmm',\n",
    " [0.1, 0.0, 0.1, 0.010, 0.005],\n",
    " [None, None, None, None, None],\n",
    " None,\n",
    " 'bound_mu_NoPriors']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.1, 0.0, 0.1, 0.01, 0.005]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "p = parameters[0]\n",
    "\n",
    "prior = p[0]\n",
    "params = p[1]\n",
    "\n",
    "print(params)\n",
    "\n",
    "hyperpriors = []\n",
    "hyperparams = p[3]\n",
    "run_description = p[4]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13866 GeneInfo(name='C3AR1', ensembl_id='ENSG00000171860.4', chrom=12, start=8210898, end=8219066)\n",
      "Found 3603 SNPs, reduced to 74 SNPs (max p-value 0.000975978) for C3AR1\n",
      "Loaded 2 features\n",
      "[-2.19722458e+00  0.00000000e+00  0.00000000e+00  1.00000000e-01\n",
      "  1.00000000e-02  4.00000000e+04]\n",
      "Starting first optimization ==============\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19722458  0.        ] 0.0 0.10000000000000002 0.010000000000000004 39999.99999999997\n",
      "0.0 0.98\n",
      "zstates.py: Working with 2 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19722458  0.        ] 0.0 0.10000000000000002 0.010000000000000004 39999.99999999997\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19722598e+00 -1.33155983e-06] 0.0 0.10000192592358798 0.010076390132891976 5413.724845791061\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19722598e+00 -1.33155983e-06] 0.0 0.10000192592358798 0.010076390132891976 5413.724845791061\n",
      "0.0 0.98\n",
      "zstates.py: Working with 2 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19722782e+00 -3.08613747e-06] 0.0 0.1000030939176877 0.010110269273876947 3940.3976209632056\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19722782e+00 -3.08613747e-06] 0.0 0.1000030939176877 0.010110269273876947 3940.3976209632056\n",
      "0.0 0.98\n",
      "zstates.py: Working with 2 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19723884e+00 -1.35583321e-05] 0.0 0.1000088494298132 0.010250010474858429 1667.3221193351017\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19723884e+00 -1.35583321e-05] 0.0 0.1000088494298132 0.010250010474858429 1667.3221193351017\n",
      "0.0 0.98\n",
      "zstates.py: Working with 2 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19725762e+00 -3.14075974e-05] 0.0 0.1000166797211524 0.010410824520763239 875.6515411021228\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19725762e+00 -3.14075974e-05] 0.0 0.1000166797211524 0.010410824520763239 875.6515411021228\n",
      "1.607254894587546e-210 0.98\n",
      "zstates.py: Working with 2 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19730139e+00 -7.30179187e-05] 0.0 0.10003038516542248 0.01066751859711796 421.8200763081295\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19730139e+00 -7.30179187e-05] 0.0 0.10003038516542248 0.01066751859711796 421.8200763081295\n",
      "1.1548794594253842e-171 0.98\n",
      "zstates.py: Working with 2 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19738747e+00 -1.54836448e-04] 0.0 0.10004830641039007 0.011006957656575589 209.30586250912683\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19738747e+00 -1.54836448e-04] 0.0 0.10004830641039007 0.011006957656575589 209.30586250912683\n",
      "1.0900812095780544e-146 0.98\n",
      "zstates.py: Working with 2 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19756612e+00 -3.22898557e-04] 0.0 0.10014115150160484 0.011256756051268432 101.41223302428615\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19756612e+00 -3.22898557e-04] 0.0 0.10014115150160484 0.011256756051268432 101.41223302428615\n",
      "6.299102927762611e-127 0.98\n",
      "zstates.py: Working with 2 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19791458e+00 -6.50743817e-04] 0.0 0.10034050621718289 0.011491226026313833 50.38497400896214\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19791458e+00 -6.50743817e-04] 0.0 0.10034050621718289 0.011491226026313833 50.38497400896214\n",
      "1.306185705791599e-108 0.98\n",
      "zstates.py: Working with 3 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19860890e+00 -1.30400168e-03] 0.0 0.10076546484904854 0.011701081770806261 25.223925441243573\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19860890e+00 -1.30400168e-03] 0.0 0.10076546484904854 0.011701081770806261 25.223925441243573\n",
      "1.5024821783692554e-90 0.98\n",
      "zstates.py: Working with 3 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.19995103 -0.00256678] 0.0 0.10161269141432698 0.011872621527806987 13.051978690700604\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.19995103 -0.00256678] 0.0 0.10161269141432698 0.011872621527806987 13.051978690700604\n",
      "2.465424231944076e-73 0.98\n",
      "zstates.py: Working with 3 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.20248005 -0.00494621] 0.0 0.10320054705261712 0.01200871538018024 7.137804640298005\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.20248005 -0.00494621] 0.0 0.10320054705261712 0.01200871538018024 7.137804640298005\n",
      "7.445658879569937e-58 0.98\n",
      "zstates.py: Working with 3 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.20698765 -0.00918691] 0.0 0.10593178841122845 0.01212007653807167 4.293587684967106\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.20698765 -0.00918691] 0.0 0.10593178841122845 0.01212007653807167 4.293587684967106\n",
      "1.5050816353409254e-45 0.98\n",
      "zstates.py: Working with 3 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.21429496 -0.0160607 ] 0.0 0.11013219461673844 0.012219047907603583 2.974128061865884\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.21429496 -0.0160607 ] 0.0 0.11013219461673844 0.012219047907603583 2.974128061865884\n",
      "1.8206175542125946e-37 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.22450602 -0.02566404] 0.0 0.11571445304622695 0.012311680299318927 2.430085654198532\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.22450602 -0.02566404] 0.0 0.11571445304622695 0.012311680299318927 2.430085654198532\n",
      "1.5886289358095427e-33 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.23731574 -0.0377081 ] 0.0 0.12253150798424073 0.012402552184139527 2.237185293729656\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.23731574 -0.0377081 ] 0.0 0.12253150798424073 0.012402552184139527 2.237185293729656\n",
      "3.2339643681643464e-32 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.2637494  -0.06255482] 0.0 0.13708712631267736 0.012567863447758294 2.0986162990894495\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.2637494  -0.06255482] 0.0 0.13708712631267736 0.012567863447758294 2.0986162990894495\n",
      "1.4323700722127995e-31 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.33153807 -0.12624857] 0.0 0.180065593036429 0.0129526480903343 1.9965724456453267\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.33153807 -0.12624857] 0.0 0.180065593036429 0.0129526480903343 1.9965724456453267\n",
      "1.6354861164748173e-31 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.45326421 -0.24045747] 0.0 0.2751405890069782 0.013550486571713795 2.089189175645669\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.45326421 -0.24045747] 0.0 0.2751405890069782 0.013550486571713795 2.089189175645669\n",
      "2.359728442163202e-32 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.61895891 -0.39523544] 0.0 0.41733352918122885 0.0142253025843121 2.3013872082882014\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.61895891 -0.39523544] 0.0 0.41733352918122885 0.0142253025843121 2.3013872082882014\n",
      "8.288449213116903e-33 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.73592867 -0.50438913] 0.0 0.5563925436877384 0.014713837373444779 2.3809916857258835\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.73592867 -0.50438913] 0.0 0.5563925436877384 0.014713837373444779 2.3809916857258835\n",
      "3.356588059997051e-32 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-2.85042591 -0.61101569] 0.0 0.728282047168559 0.015190969316937324 2.382462001723677\n",
      "called from prob_comps\n",
      "Before czcomps: [-2.85042591 -0.61101569] 0.0 0.728282047168559 0.015190969316937324 2.382462001723677\n",
      "5.225146583383123e-31 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.18222022 -0.90496721] 0.0 0.6861888732786208 0.015548943959667267 2.1846850228605708\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.18222022 -0.90496721] 0.0 0.6861888732786208 0.015548943959667267 2.1846850228605708\n",
      "1.0711367527453028e-28 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.14075469 -0.87109273] 0.0 0.8107971464680316 0.0157115405229485 2.2888599669387846\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.14075469 -0.87109273] 0.0 0.8107971464680316 0.0157115405229485 2.2888599669387846\n",
      "5.0369038912566814e-29 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.17450373 -0.90177856] 0.0 0.8431118920182915 0.015809559142339478 2.3132867291496\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.17450373 -0.90177856] 0.0 0.8431118920182915 0.015809559142339478 2.3132867291496\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6.307861064037395e-29 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.23696704 -0.95721384] 0.0 0.8432638989017303 0.01589809950155548 2.3295419962996373\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.23696704 -0.95721384] 0.0 0.8432638989017303 0.01589809950155548 2.3295419962996373\n",
      "8.844827499152033e-29 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.29227955 -1.00510641] 0.0 0.8002935318947753 0.01591233006672515 2.3340847277399863\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.29227955 -1.00510641] 0.0 0.8002935318947753 0.01591233006672515 2.3340847277399863\n",
      "9.802759653599849e-29 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.31609013 -1.02246536] 0.0 0.7433711127768056 0.015885016607460507 2.328467344139442\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.31609013 -1.02246536] 0.0 0.7433711127768056 0.015885016607460507 2.328467344139442\n",
      "9.554548887772152e-29 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.31221622 -1.01739624] 0.0 0.7376760650321555 0.015889577354701867 2.3313388685422285\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.31221622 -1.01739624] 0.0 0.7376760650321555 0.015889577354701867 2.3313388685422285\n",
      "9.320447714626882e-29 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.30692748 -1.01122153] 0.0 0.7346233884793882 0.015895313393279816 2.3325167671162164\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.30692748 -1.01122153] 0.0 0.7346233884793882 0.015895313393279816 2.3325167671162164\n",
      "9.331893196141209e-29 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.30552204 -1.00878535] 0.0 0.733934894663715 0.01590706908491602 2.3325627904594346\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.30552204 -1.00878535] 0.0 0.733934894663715 0.01590706908491602 2.3325627904594346\n",
      "9.846916191144022e-29 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.32903318 -0.98238151] 0.0 0.7300140979168703 0.01647164838560103 2.342987685419097\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.32903318 -0.98238151] 0.0 0.7300140979168703 0.01647164838560103 2.342987685419097\n",
      "1.302858246031783e-27 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.32652259 -0.97971712] 0.0 0.7306041108110952 0.016473466713401753 2.3370704022266753\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.32652259 -0.97971712] 0.0 0.7306041108110952 0.016473466713401753 2.3370704022266753\n",
      "1.4071598652141284e-27 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.33386511 -0.97624609] 0.0 0.7329527561150688 0.016598250756307276 2.334323760467851\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.33386511 -0.97624609] 0.0 0.7329527561150688 0.016598250756307276 2.334323760467851\n",
      "2.6113543211472535e-27 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.33782844 -0.97540682] 0.0 0.7337467814805614 0.016650035615725138 2.334259670950572\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.33782844 -0.97540682] 0.0 0.7337467814805614 0.016650035615725138 2.334259670950572\n",
      "3.328342288398245e-27 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.34150701 -0.97312842] 0.0 0.7342735124013658 0.01666980358159262 2.3342201238263693\n",
      "called from func_grad\n",
      "Before czcomps: [-3.35622131 -0.96401482] 0.0 0.7363842200153587 0.016749110421556662 2.334061942029536\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.35622131 -0.96401482] 0.0 0.7363842200153587 0.016749110421556662 2.334061942029536\n",
      "5.28666711629579e-27 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.45337444 -0.88559185] 0.0 0.7437545613679968 0.017059407488757332 2.3337010416541637\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.45337444 -0.88559185] 0.0 0.7437545613679968 0.017059407488757332 2.3337010416541637\n",
      "2.1757409177247613e-26 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-3.70598983 -0.65673226] 0.0 0.7559840503721352 0.017484931444083716 2.3327610797921965\n",
      "called from prob_comps\n",
      "Before czcomps: [-3.70598983 -0.65673226] 0.0 0.7559840503721352 0.017484931444083716 2.3327610797921965\n",
      "1.4207659530057535e-25 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-4.3154175  -0.06622742] 0.0 0.7662800996843857 0.01803974847035591 2.33226545657378\n",
      "called from prob_comps\n",
      "Before czcomps: [-4.3154175  -0.06622742] 0.0 0.7662800996843857 0.01803974847035591 2.33226545657378\n",
      "1.4391082761592538e-24 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-5.77789534  1.38244553] 0.0 0.8187848450061564 0.017668071572571196 2.3167965872594976\n",
      "called from prob_comps\n",
      "Before czcomps: [-5.77789534  1.38244553] 0.0 0.8187848450061564 0.017668071572571196 2.3167965872594976\n",
      "3.8166888346551863e-25 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-9.15799813  4.9049537 ] 0.0 0.8406354067484868 0.018864969860654536 2.33233132286228\n",
      "called from prob_comps\n",
      "Before czcomps: [-9.15799813  4.9049537 ] 0.0 0.8406354067484868 0.018864969860654536 2.33233132286228\n",
      "3.126653014545926e-23 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-19.53852385  15.7830937 ] 0.0 0.9533290163056233 0.019020239961277993 2.3370089698192142\n",
      "called from prob_comps\n",
      "Before czcomps: [-19.53852385  15.7830937 ] 0.0 0.9533290163056233 0.019020239961277993 2.3370089698192142\n",
      "3.3731345161065063e-23 0.98\n",
      "zstates.py: Working with 5 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-73.26363057  69.88002991] 0.0 0.6916660148838937 0.012775260905603597 2.4610198799731933\n",
      "called from prob_comps\n",
      "Before czcomps: [-73.26363057  69.88002991] 0.0 0.6916660148838937 0.012775260905603597 2.4610198799731933\n",
      "4.15848032996772e-38 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-92.18621067  89.59907481] 0.0 0.8122672969751698 0.012621541923701256 2.3777178788022444\n",
      "called from prob_comps\n",
      "Before czcomps: [-92.18621067  89.59907481] 0.0 0.8122672969751698 0.012621541923701256 2.3777178788022444\n",
      "3.0988005462639694e-38 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-104.82429058  102.26805487] 0.0 0.6775381490274182 0.012466562646535053 2.3127208601338336\n",
      "called from prob_comps\n",
      "Before czcomps: [-104.82429058  102.26805487] 0.0 0.6775381490274182 0.012466562646535053 2.3127208601338336\n",
      "3.679397587889585e-38 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-120.36881092  117.95118406] 0.0 0.6534265551992728 0.012907518649776824 2.305913280482888\n",
      "called from prob_comps\n",
      "Before czcomps: [-120.36881092  117.95118406] 0.0 0.6534265551992728 0.012907518649776824 2.305913280482888\n",
      "7.567013472627867e-37 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-199.45184748  197.47298642] 0.0 0.6757625670945111 0.01667059906577939 2.299533205958446\n",
      "called from prob_comps\n",
      "Before czcomps: [-199.45184748  197.47298642] 0.0 0.6757625670945111 0.01667059906577939 2.299533205958446\n",
      "5.8377430293425425e-28 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-274.81852847  273.01864594] 0.0 0.6567436862618592 0.019107545914974006 2.329960664780146\n",
      "called from prob_comps\n",
      "Before czcomps: [-274.81852847  273.01864594] 0.0 0.6567436862618592 0.019107545914974006 2.329960664780146\n",
      "7.2824801743989e-24 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-399.69518171  398.23813925] 0.0 0.719159458673184 0.018731152834424218 2.3494810160743116\n",
      "called from prob_comps\n",
      "Before czcomps: [-399.69518171  398.23813925] 0.0 0.719159458673184 0.018731152834424218 2.3494810160743116\n",
      "1.0871638969156604e-24 0.98\n",
      "zstates.py: Working with 4 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-600.10523735  598.92059625] 0.0 0.7507952043185241 0.016933676881760586 2.3514103151515933\n",
      "called from prob_comps\n",
      "Before czcomps: [-600.10523735  598.92059625] 0.0 0.7507952043185241 0.016933676881760586 2.3514103151515933\n",
      "5.149596213612865e-28 0.98\n",
      "zstates.py: Working with 3 leading zstates.\n",
      "called from func_grad\n",
      "Before czcomps: [-871.31376657  870.63315942] 0.0 0.7666960879032239 0.016446019552374534 2.3433786530614773\n",
      "called from func_grad\n",
      "Before czcomps: [-607.57969635  606.40894634] 0.0 0.7512289790794705 0.016920045304900554 2.3511885955264864\n",
      "called from func_grad\n",
      "Before czcomps: [-601.11231473  599.92954526] 0.0 0.7508536347142833 0.016931839580071856 2.3513804403568486\n",
      "called from func_grad\n",
      "Before czcomps: [-600.12122841  598.93661703] 0.0 0.7507961320802443 0.016933647706287042 2.35140984077648\n",
      "called from func_grad\n",
      "Before czcomps: [-600.10524132  598.92060023] 0.0 0.750795204548532 0.016933676874527473 2.351410315033988\n",
      "called from func_grad\n",
      "Before czcomps: [-600.10523735  598.92059625] 0.0 0.7507952043185241 0.016933676881760586 2.3514103151515933\n",
      "called from func_grad\n",
      "Before czcomps: [-600.10523735  598.92059625] 0.0 0.7507952043185241 0.016933676881760586 2.3514103151515933\n",
      "called from func_grad\n",
      "Before czcomps: [-600.10523735  598.92059625] 0.0 0.7507952043185241 0.016933676881760586 2.3514103151515933\n",
      "called from prob_comps\n",
      "Before czcomps: [-600.10523735  598.92059625] 0.0 0.7507952043185241 0.016933676881760586 2.3514103151515933\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.149596213612865e-28 0.98\n",
      "zstates.py: Working with 3 leading zstates.\n",
      "      fun: 344.81271448497097\n",
      " hess_inv: <6x6 LbfgsInvHessProduct with dtype=float64>\n",
      "      jac: array([ 0.09153588,  0.09026455, -0.01298836,  0.04184904,  0.05923147,\n",
      "       -1.92146256])\n",
      "  message: b'CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH'\n",
      "     nfev: 59\n",
      "      nit: 50\n",
      "   status: 0\n",
      "  success: True\n",
      "        x: array([-6.00105237e+02,  5.98920596e+02,  0.00000000e+00, -2.86622362e-01,\n",
      "       -4.07845092e+00, -4.27507641e-01])\n",
      "called from model_exp\n",
      "Before czcomps: [-600.10523735  598.92059625] 0.0 0.7507952043185241 0.016933676881760586 2.3514103151515933\n"
     ]
    }
   ],
   "source": [
    "%load_ext wurlitzer\n",
    "from iotools import snp_annotator\n",
    "\n",
    "model_dir = prior+\"_{:.3f}_{:.3f}_{:.3f}_{:.3f}_{:.3f}\".format(params[0], params[1], params[2], params[3], params[4])\n",
    "modelpath = os.path.join(\"./z\"+str(zmax), run_description, model_dir)\n",
    "\n",
    "write_params(modelpath, p)\n",
    "\n",
    "model = WriteModel(modelpath, chrom)\n",
    "\n",
    "# for i in range(0,len(gene_training_list)):\n",
    "for i in range(0,1):\n",
    "\n",
    "    k, gene = gene_training_list[i]\n",
    "\n",
    "    print(k, gene)\n",
    "\n",
    "    # select only the cis-SNPs\n",
    "    cismask = mfunc.select_snps(gene, f_snps, window)\n",
    "    if len(cismask) > 0:\n",
    "        target = expression[k, exprmask]\n",
    "        target = scale(target, with_mean=True, with_std=True)\n",
    "        predictor = gt[cismask][:, vcfmask]\n",
    "        snpmask = cismask\n",
    "\n",
    "        # if number of cis SNPs > threshold, use p-value cut-off\n",
    "        min_snps = 20\n",
    "        if len(cismask) > min_snps:\n",
    "            assoc_model = LinRegAssociation(predictor, target, min_snps, pval_cutoff)\n",
    "            pvalmask = cismask[assoc_model.selected_variables]\n",
    "            print (\"Found {:d} SNPs, reduced to {:d} SNPs (max p-value {:g}) for {:s}\".format(len(cismask), len(pvalmask), assoc_model.ordered_pvals[len(pvalmask) - 1], gene.name))\n",
    "            predictor = gt[pvalmask][:, vcfmask]\n",
    "            snpmask = pvalmask\n",
    "        else:\n",
    "            print (\"Found {:d} SNPs for {:s}\".format(len(cismask), gene.name))\n",
    "\n",
    "            \n",
    "        # read the features\n",
    "        # TO-DO: call another module for getting the features\n",
    "        # for now, it contains only a list of 1's\n",
    "        feature1 = np.ones((predictor.shape[0], 1))\n",
    "        \n",
    "        selected_snps = [f_snps[x] for x in snpmask]\n",
    "        feature2 = snp_annotator.get_dummy_dist_feature(selected_snps, gene, window)\n",
    "        # feature2 = (feature2 - np.mean(feature2))/np.std(feature2)\n",
    "        \n",
    "#         np.random.shuffle(feature2)\n",
    "      \n",
    "        \n",
    "#         feature3 = np.random.random_integers(0,1, len(selected_snps))[np.newaxis].T\n",
    "        feature3 = np.random.random(len(selected_snps))[np.newaxis].T\n",
    "    \n",
    "        feature4 = np.random.exponential(scale=1, size=len(selected_snps))[np.newaxis].T\n",
    "        \n",
    "        \n",
    "        features = np.concatenate((feature1,feature2), axis=1)\n",
    "        \n",
    "        nfeat = features.shape[1]\n",
    "        print(\"Loaded {:d} features\".format(nfeat))\n",
    "\n",
    "        init_params = np.zeros(nfeat + 4)\n",
    "        init_params[0] = - np.log((1 / params[0]) - 1)\n",
    "        for i in range(1, nfeat):\n",
    "            init_params[i] = 0\n",
    "            # init_params[i] = params[0]\n",
    "        init_params[nfeat + 0] = params[1] # mu\n",
    "        init_params[nfeat + 1] = params[2] # sigma\n",
    "        init_params[nfeat + 2] = params[3] # sigmabg\n",
    "        init_params[nfeat + 3] = 1 / params[4] / params[4] # tau\n",
    "\n",
    "        print(init_params)          \n",
    "            \n",
    "        # perform the analysis\n",
    "\n",
    "        print (\"Starting first optimization ==============\")\n",
    "        emp_bayes = EmpiricalBayes(predictor, target, features, 1, init_params, method=\"new\")\n",
    "        emp_bayes.fit()\n",
    "        if zmax > 1:\n",
    "            if emp_bayes.success:\n",
    "                res = emp_bayes.params\n",
    "                print (\"Starting second optimization from previous results ================\")\n",
    "                # Python Error: C library could not compute z-components. Check C errors above.\n",
    "            else:\n",
    "                res = init_params\n",
    "                print (\"Starting second optimization from initial parameters ================\")\n",
    "            emp_bayes = EmpiricalBayes(predictor, target, features, cmax, res, method=\"new\")\n",
    "            emp_bayes.fit()\n",
    "\n",
    "        if emp_bayes.success:\n",
    "            res = emp_bayes.params\n",
    "#             res[4] = 1 / np.sqrt(res[4])\n",
    "#             print(\"PI: \\t\",res[0])\n",
    "#             print(\"mu: \\t\",res[1])\n",
    "#             print(\"sigma: \\t\",res[2])\n",
    "#             print(\"sigmabg: \\t\",res[3])\n",
    "#             print(\"tau: \\t\",res[4])\n",
    "\n",
    "            model_snps = [f_snps[x] for x in snpmask]\n",
    "            model_zstates = list()\n",
    "            scaledparams = hyperparameters.scale(emp_bayes.params)\n",
    "            zprob, zexp = logmarglik.model_exp(scaledparams, predictor, target, features, emp_bayes.zstates)\n",
    "            for j, z in enumerate(emp_bayes.zstates):\n",
    "                this_zstate = ZstateInfo(state = z,\n",
    "                                         prob  = zprob[j],\n",
    "                                         exp   = list(zexp[j, :]) )\n",
    "                model_zstates.append(this_zstate)\n",
    "            # print(model_snps)\n",
    "            # for i,m in enumerate(model_zstates):\n",
    "            #     print(\"z-state: \",i,\" Prob:\", m.prob)\n",
    "            model.write_success_gene(gene, model_snps, model_zstates, res)\n",
    "        else:\n",
    "            model.write_failed_gene(gene, res) # np.zeros_like(init_params)\n",
    "            print (\"Failed optimization\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "sys.path.append(\"../\")\n",
    "import os\n",
    "import pickle\n",
    "from utils.printstamp import printStamp\n",
    "from iotools.io_model import ReadModel\n",
    "from utils.containers import GeneExpressionArray\n",
    "from utils import gtutils\n",
    "from utils import mfunc\n",
    "import numpy as np\n",
    "from config_annots import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-03-14 20:33:59 - Reading pickled genotype\n",
      "2018-03-14 20:34:06 - Done reading\n"
     ]
    }
   ],
   "source": [
    "\n",
    "if not os.path.exists(p_pickfile):\n",
    "# Read genotype (quite slow for testing) use pickle below\n",
    "    p_oxf = ReadOxford(p_gtpath, p_samplepath, chrom, predicting_dataset)\n",
    "    p_genotype = np.array(p_oxf.dosage)\n",
    "    p_samplenames = p_oxf.samplenames\n",
    "    p_snps = p_oxf.snps_info\n",
    "    p_nsample = len(p_oxf.samplenames)\n",
    "\n",
    "    printStamp(\"Dumping CHR {:d} genotype\".format(chrom))\n",
    "    with open(p_pickfile, 'wb') as output:\n",
    "        pickle.dump(p_oxf, output, pickle.HIGHEST_PROTOCOL)\n",
    "else:\n",
    "    printStamp(\"Reading pickled genotype\")\n",
    "    with open(p_pickfile, 'rb') as input:\n",
    "        pickled_oxf = pickle.load(input)\n",
    "\n",
    "    printStamp(\"Done reading\")\n",
    "\n",
    "    p_genotype = np.array(pickled_oxf.dosage)\n",
    "    p_samplenames = pickled_oxf.samplenames\n",
    "    p_snps = pickled_oxf.snps_info\n",
    "    p_nsample = len(pickled_oxf.samplenames)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-03-14 20:34:22 - Predicting for gxpred-bslmm_0.100_0.000_0.100_0.010_0.005\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Use parameters from config.py\n",
    "# for p in parameters:\n",
    "p = parameters[0]\n",
    "prior = p[0]\n",
    "params = p[1]\n",
    "hyperpriors = p[2]\n",
    "hyperparams = p[3]\n",
    "run_description = p[4]\n",
    "model_dir = prior+\"_{:.3f}_{:.3f}_{:.3f}_{:.3f}_{:.3f}\".format(params[0], params[1], params[2], params[3], params[4])\n",
    "modelpath = os.path.join(\"./z\"+str(zmax)+ \"_test\", run_description, model_dir)\n",
    "\n",
    "outfileprefix = os.path.join(modelpath,\"pred_chr\"+str(chrom))\n",
    "\n",
    "\n",
    "printStamp(\"Predicting for \"+model_dir)\n",
    "# Write predictions for each model\n",
    "p_model = ReadModel(modelpath, chrom)\n",
    "p_genes = p_model.genes\n",
    "gx = list()\n",
    "for gene in p_genes:\n",
    "\n",
    "    p_model.read_gene(gene)\n",
    "    p_model_snps = p_model.snps\n",
    "    p_model_zstates = p_model.zstates\n",
    "\n",
    "    x = gtutils.prediction_variables(p_snps, p_model_snps, p_genotype)\n",
    "    x = gtutils.normalize(p_model_snps, x)\n",
    "\n",
    "    ypred = np.zeros(p_nsample)\n",
    "    for z in p_model_zstates:\n",
    "        ypred += z.prob * np.dot(x.T, z.exp)\n",
    "\n",
    "    gx.append(GeneExpressionArray(geneid = gene.ensembl_id, expr_arr = ypred))\n",
    "\n",
    "\n",
    "# Write output\n",
    "mfunc.write_gcta_phenotype(outfileprefix, p_samplenames, gx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assessment "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../\")\n",
    "import os\n",
    "from iotools import readgtf\n",
    "from iotools.readrpkm import ReadRPKM\n",
    "from iotools.readPrediction import ReadPrediction\n",
    "from scipy.stats import pearsonr\n",
    "from helper_functions import load_target_genes, write_r2_dataframe, get_common_elements, new_write_predicted_r2, pearson_corr_rowwise\n",
    "import math\n",
    "import pickle\n",
    "from utils.printstamp import printStamp\n",
    "\n",
    "from config_annots import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Load reference dataset Gene Expression\n",
    "reference_rpkm = ReadRPKM(reference_expdatapath, \"cardiogenics\")\n",
    "reference_expression = reference_rpkm.expression\n",
    "reference_expr_donors = reference_rpkm.donor_ids\n",
    "reference_gene_names = reference_rpkm.gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 70 genes with high r2 values\n",
      "\n",
      "Found 7 genes in CHR 12\n"
     ]
    }
   ],
   "source": [
    "# use the selected_gene_ids with high RÂ² values as targets, only those in the selected chrom will appear\n",
    "genelistfile = \"genes4testing_highr2\"\n",
    "selected_gene_ids = load_target_genes(genelistfile, gene_info, chrom)\n",
    "target_genelist = [g.split(\".\")[0] for g in selected_gene_ids]\n",
    "target_donors = reference_expr_donors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2018-03-19 15:16:10 - Reading pickled Predixcan prediction\n",
      "Genes found: 7 of 7\n",
      "Samples found: 744 of 849\n"
     ]
    }
   ],
   "source": [
    "\n",
    "### Predixcan assessment ###\n",
    "\n",
    "if not os.path.exists(predixcan_pickfile):\n",
    "    # pxpred_predpath = os.path.join(home, \"predictions/cardiogenics/predixcan_predictions_klinikum\")\n",
    "    predixcanpred = ReadPrediction(pxpred_predpath, reference_samplepath, \"predixcan\", trim=True)\n",
    "\n",
    "    printStamp(\"Dumping Predixcan prediction\")\n",
    "    with open(predixcan_pickfile, 'wb') as output:\n",
    "        pickle.dump(predixcanpred, output, pickle.HIGHEST_PROTOCOL)\n",
    "else:\n",
    "    printStamp(\"Reading pickled Predixcan prediction\")\n",
    "    with open(predixcan_pickfile, 'rb') as input:\n",
    "        predixcanpred = pickle.load(input)\n",
    "\n",
    "# filter predixcan predictions with only those in gxpred\n",
    "predixcanpred.sort_by_gene(target_genelist)\n",
    "predixcanpred.sort_by_samples(target_donors, use_prev=True)\n",
    "\n",
    "sorted_expr_donors, ix_samples = get_common_elements(reference_expr_donors, predixcanpred.sorted_samples)\n",
    "sorted_gene_names, ix_genes = get_common_elements(reference_gene_names, predixcanpred.sorted_gene_names)\n",
    "sorted_expression = reference_expression[ix_genes,:][:, ix_samples].T\n",
    "\n",
    "predixcan_r = pearson_corr_rowwise(predixcanpred.sorted_expr_mat.T, sorted_expression.T)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "gxpred-bslmm [0.1, 0.0, 0.1, 0.01, 0.005]\n",
      "./z1_test/bound_mu_NoPriors/gxpred-bslmm_0.100_0.000_0.100_0.010_0.005\n",
      "No prediction found for CHR 1\n",
      "No prediction found for CHR 2\n",
      "No prediction found for CHR 3\n",
      "No prediction found for CHR 4\n",
      "No prediction found for CHR 5\n",
      "No prediction found for CHR 6\n",
      "No prediction found for CHR 7\n",
      "No prediction found for CHR 8\n",
      "No prediction found for CHR 9\n",
      "No prediction found for CHR 10\n",
      "No prediction found for CHR 11\n",
      "Loaded 1 genes in CHR 12\n",
      "1\n",
      "No prediction found for CHR 13\n",
      "No prediction found for CHR 14\n",
      "No prediction found for CHR 15\n",
      "No prediction found for CHR 16\n",
      "No prediction found for CHR 17\n",
      "No prediction found for CHR 18\n",
      "No prediction found for CHR 19\n",
      "No prediction found for CHR 20\n",
      "No prediction found for CHR 21\n",
      "No prediction found for CHR 22\n",
      "Genes found: 1 of 7\n",
      "Samples found: 744 of 849\n",
      "(786, 1)\n",
      "['ENSG00000171860']\n",
      "[0.60532227]\n",
      "[0.60396525 0.28503124 0.63198656 0.62872473 0.50988101 0.21494123\n",
      " 0.19620032]\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "### GXpred assessment ###\n",
    "\n",
    "#for p in parameters:\n",
    "p = parameters[0]\n",
    "prior = p[0]\n",
    "params = p[1]\n",
    "hyperpriors = p[2]\n",
    "hyperparams = p[3]\n",
    "run_description = p[4]\n",
    "model_dir = prior+\"_{:.3f}_{:.3f}_{:.3f}_{:.3f}_{:.3f}\".format(params[0], params[1], params[2], params[3], params[4])\n",
    "print(prior, params)\n",
    "\n",
    "modelpath = os.path.join(\"./z\"+str(zmax)+ \"_test\", run_description, model_dir)\n",
    "print(modelpath)\n",
    "\n",
    "\n",
    "gxpred_predpath = os.path.join(modelpath)\n",
    "gxpred = ReadPrediction(gxpred_predpath, reference_samplepath, \"gxpred\", trim=True)\n",
    "\n",
    "# filter gxpred predicted values\n",
    "gxpred.sort_by_gene(target_genelist)\n",
    "gxpred.sort_by_samples(target_donors, use_prev=True)\n",
    "\n",
    "\n",
    "# Filter and sort the reference expression values\n",
    "# Cardiogenics variables\n",
    "# expression\n",
    "# expr_donors\n",
    "# gene_names\n",
    "\n",
    "sorted_expr_donors, ix_samples = get_common_elements(reference_expr_donors, gxpred.sorted_samples)\n",
    "sorted_gene_names, ix_genes = get_common_elements(reference_gene_names, gxpred.sorted_gene_names)\n",
    "sorted_expression = reference_expression[ix_genes,:][:, ix_samples].T\n",
    "\n",
    "# Calculate Pearson correlation\n",
    "gxpred_r = pearson_corr_rowwise(gxpred.sorted_expr_mat.T, sorted_expression.T)\n",
    "\n",
    "print(gxpred.sorted_gene_names)\n",
    "print(gxpred_r**2)\n",
    "print(predixcan_r**2)\n",
    "\n",
    "\n",
    "# Write to table with predictions for given genes\n",
    "# predtabledir = os.path.join(home, \"gxpred\",\"devtools\", \"all_predictions.txt\")\n",
    "# new_write_predicted_r2(predtabledir, prior, params, gxpred_r, predixcan_r, gxpred.sorted_gene_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes_df:                       Ensembl_ID Gene_Name  Success     Gamma0    Gamma1   Mu  \\\n",
      "ENSG00000171860  ENSG00000171860     C3AR1     True -600.10524  598.9206  0.0   \n",
      "\n",
      "                  Sigma  Sigma_bg  Sigma_tau  \n",
      "ENSG00000171860  0.7508   0.01693    2.35141  \n",
      "new_df:                  predixcan       Ensembl_ID\n",
      "ENSG00000171860   0.603965  ENSG00000171860\n",
      "ENSG00000064115   0.285031  ENSG00000064115\n",
      "ENSG00000139610   0.631987  ENSG00000139610\n",
      "ENSG00000090382   0.628725  ENSG00000090382\n",
      "ENSG00000127337   0.509881  ENSG00000127337\n",
      "ENSG00000135643   0.214941  ENSG00000135643\n",
      "ENSG00000184752   0.196200  ENSG00000184752\n",
      "(1, 9)\n",
      "(7, 2)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "write_r2_dataframe(modelpath, chrom, \"predixcan\", predixcan_r, predixcanpred)\n",
    "write_r2_dataframe(modelpath, chrom, prior, gxpred_r, gxpred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
